src/classes/COS_StartEndTime_TDTM.cls

```diff
@@ -0,0 +1,194 @@
+        // and to flow down the times to the Course Offering Schedule if necessary. We'll also do this if the Time Block lookup field is swapped.
+        for (Course_Offering_Schedule__c courseOfferingSchedule : courseOfferingSchedules) {
+            
+            // We don't need to flow down any times or do any validation if there is no Time Block.
+            if (courseOfferingSchedule.Time_Block__c == null) {
+                continue;
+            }
+
+            timeBlockIdsToQuery.add(courseOfferingSchedule.Time_Block__c);
+            courseOfferingSchedulesToValidate.add(courseOfferingSchedule);
+
+            if (isUpdateContext) {
+                Course_Offering_Schedule__c oldCourseOfferingSchedule = oldCourseOfferingSchedulesMap.get(courseOfferingSchedule.Id);
+
+                // Is the Time Block changing? We'll need to grab the Start and End Times from both Time Block records before we update the
+                // Start and End Times on the Course Offering Schedule.
+                if (oldCourseOfferingSchedule != null && oldCourseOfferingSchedule.Time_Block__c != courseOfferingSchedule.Time_Block__c) {
+                    timeBlockIdsToQuery.add(oldCourseOfferingSchedule.Time_Block__c);
+                }
+            }
+        }
+
+        // Exit early if there are no Course Offering Schedules to validate.
+        if (courseOfferingSchedulesToValidate.size() == 0) {
+            return dmlWrapper;
+        }
+
+        // Query the Time Blocks to verify the times are valid and flow down the times to the Course Offering Schedules if necessary.
+        validateCourseOfferingSchedules(timeBlockIdsToQuery, courseOfferingSchedulesToValidate, oldCourseOfferingSchedulesMap, isUpdateContext);
+
+        return dmlWrapper;
+    }
+
+    /*******************************************************************************************************
+    * @description Queries for Time Blocks related to the Course Offering Schedules to validate. The times
+    * from these Time Block records are then flowed down to the Course Offering Schedules. An error is thrown
+    * if the times on the Course Offering Schedule do not match the times on the Time Block.
+    * @param timeBlockIdsToQuery The Set of Time Block IDs that should be queried.
+    * @param courseOfferingSchedulesToValidate The list of Course Offering Schedule records to validate and
+    * flow times down to.
+    * @param oldCourseOfferingSchedulesMap The map of old Course Offering Schedule records prior to the update,
+    * if in an update context.
+    * @param isUpdateContext True if the trigger is currently executing in a BeforeUpdate context.
+    ********************************************************************************************************/
+    private void validateCourseOfferingSchedules(Set<Id> timeBlockIdsToQuery, List<Course_Offering_Schedule__c> courseOfferingSchedulesToValidate, 
```

Comment:
Yeah, I agree - because the code is similar, I chose to combine it instead of having it duplicated. I've made a few adjustments in my upcoming commit that hopefully makes this a bit easier to read.